src/os/bluestore/BlueFS.cc

```diff
@@ -1391,7 +1393,15 @@ int BlueFS::_flush_and_sync_log(std::unique_lock<std::mutex>& l,
     int r = _allocate(log_writer->file->fnode.prefer_bdev,
 		      cct->_conf->bluefs_max_log_runway,
 		      &log_writer->file->fnode.extents);
+    if ((r == -ENOSPC) && _should_compact_log() && _try_num) {
+      // log_writer->file->fnode.extents may hint the end of the dev...
+      _compact_log_sync(); 
```

Comment:
I don't think it's safe to call _compat_log_sync() here.  See the first few lines of that function:
<pre>
  // clear out log (be careful who calls us!!!)
  log_t.clear();
</pre>
which will blow away at least something (given we just asserted that log_t is not empty).

That aside, I'm not sure this makes a lot of sense or that I understand the problem anyway.  Compacting will make the log more efficient, but this allocation will only fail if *all* devices are out of space... not just BDEV_WAL.  Which means you must be running dangerously close to full...?